std-bits/table/src/main/java/org/enso/table/operations/AddRowNumber.java

```diff
@@ -0,0 +1,122 @@
+package org.enso.table.operations;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+import org.enso.base.text.TextFoldingStrategy;
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.column.storage.numeric.LongStorage;
+import org.enso.table.data.column.storage.type.IntegerType;
+import org.enso.table.data.index.MultiValueIndex;
+import org.enso.table.data.index.OrderedMultiValueKey;
+import org.enso.table.data.index.UnorderedMultiValueKey;
+import org.enso.table.data.table.Column;
+import org.enso.table.util.ConstantList;
+import org.graalvm.collections.Pair;
+
+public class AddRowNumber {
+
+  public static LongStorage create_grouped_numbering(
+      long start, long step, Column[] groupingColumns) {
+    assert groupingColumns.length > 0;
```

Comment:
I'd personally prefer doing argument checking and throwing an `IllegalArgumentException` in this kind of case. I appreciate it is an internal method but as a public method on the jar it could be used in other places eventially.

Reply:
Hmmm, I don't quite like the philosophy of working with internal code as if it may be accessed publicly.

If we have abstractions leaking, we should improve our encapsulation (which I know we are working on).

---

But ofc there is no harm in making the error messages clearer, so I can add the exception.